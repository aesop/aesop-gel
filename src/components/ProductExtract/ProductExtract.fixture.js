export default {
  dataTestRef: 'test_id',
  product: {
    image: {
      altText: 'Product Extract',
      small:
        '//images.ctfassets.net/u1nb1km7t5q7/5JHOiuwRLBAc78jTcLVb5I/bfad1667393889dd8af5e31b080436bd/Aesop_Hair_Violet_Leaf_Hair_Balm_60mL_small.png',
    },
    name: 'The sea',
    url: '/',
  },
  works: 'Works well with',
};
