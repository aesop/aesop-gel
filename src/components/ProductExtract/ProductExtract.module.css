@import 'styles/index';

.base {
  width: 100%;
  padding: 0;
  margin: 0;
  margin-top: 18px;
  list-style: none;
  visibility: hidden;

  @media (--viewport-md) {
    margin-top: 14px;
  }
}

.heading,
.productName {
  display: block;
  font-size: rem(14px);
  line-height: 1.4;

  @media (--viewport-md) {
    line-height: 1.6;
  }
}

.delayAnimation {
  opacity: 0;
  transform: translateY(-20px);
  transition-duration: 1000ms;
  transition-property: all;
  transition-timing-function: var(--easing-ease-out-cubic);
  visibility: hidden;

  .item1 & {
    transition-delay: calc((1 * 100ms) + 200ms);
  }

  .item2 & {
    transition-delay: calc((2 * 100ms) + 200ms);
  }

  .item3 & {
    transition-delay: calc((3 * 100ms) + 200ms);
  }

  .item4 & {
    transition-delay: calc((4 * 100ms) + 200ms);
  }
}

.heading {
  margin-bottom: 2px;
  color: var(--color-secondary);
  font-family: var(--font-family-sans-serif-medium);
  font-size: rem(14px);

  .dark & {
    @extend %font-theme-dark-copy;
  }

  .light & {
    color: var(--color-light-copy);
  }
}

.productName {
  position: relative;
  padding-bottom: 18px;
  margin-left: 0;
  color: var(--color-dark-label);
  font-family: var(--font-family-sans-serif-regular);

  .dark & {
    @extend %font-theme-dark-label;
  }

  .light & {
    color: var(--color-light-label);
  }

  @media (--viewport-md) {
    padding-bottom: 14px;
  }

  span {
    border-bottom: 1px solid transparent;
    transition: border-bottom 300ms var(--easing-ease-in-sine);
  }

  .link:hover & {
    span {
      border-bottom: 1px solid var(--color-grey-30);
    }
  }
}

.productImage {
  img {
    width: auto;
    max-height: 50px;
  }
}

.slideIn {
  opacity: 1;
  transform: translateY(0);
  visibility: visible;
}

.link {
  display: inline-block;
  width: 100%;
}

.flexContainer {
  display: flex;
  justify-content: space-between;
}

.hasBottomBorder {
  border-bottom: 1px solid var(--color-grey-65);
}
