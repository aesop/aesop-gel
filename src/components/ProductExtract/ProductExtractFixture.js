export default {
  dataTestRef: 'test_id',
  product: {
    works: 'Works well with',
    image: {
      altText: 'Product Extract',
      small:
        'https://www.aesop.com/medias/Aesop-Hand-Resurrection-Rinse-Free-Hand-Wash-500mL-large.png?context=bWFzdGVyfGltYWdlc3wxNjEyNDF8aW1hZ2UvcG5nfGltYWdlcy9oNTgvaDRkLzkwMTkyNTQ5OTcwMjIucG5nfGU5YTRhMDk0Y2FjZWVmYzMyN2MxYTEwZTA2NjJhNTY1MDAwZmRkNzBmOWQ2MGIzYTNiYWM0MjVmODFlOGFmYzc',
    },
    name: 'The sea',
    url: '/',
  },
};
