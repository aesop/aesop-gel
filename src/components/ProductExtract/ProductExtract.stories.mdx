import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { ProductExtract } from './ProductExtract';
import { ProductExtractFixture } from './ProductExtract.fixture';

<Meta title="Components/ProductExtract" component={ProductExtract} />

# ProductExtract

<Canvas>
  <Story name="Base component">
    <ProductExtract
      dataTestRef={ProductExtractFixture.dataTestRef}
      hasBottomBorder={knobs.boolean('hasBottomBorder', false, 'Presentation')}
      isVisible={knobs.boolean('isActive', true, 'Presentation')}
      itemNum={knobs.select('itemNum', [1, 2, 3, 4], 4, 'Presentation')}
      product={knobs.object(
        'product',
        ProductExtractFixture.product,
        'product',
      )}
      theme={knobs.select('theme', ['dark', 'light'], 'dark', 'Presentation')}
      works={ProductExtractFixture.works}
    />
  </Story>
</Canvas>

## API

<ArgsTable of={ProductExtract} />
