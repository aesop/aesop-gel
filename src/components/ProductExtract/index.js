import ProductExtract from './ProductExtract';

export default ProductExtract;
