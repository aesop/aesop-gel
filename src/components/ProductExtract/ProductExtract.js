import React, { forwardRef, Fragment } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Image from '~/components/Image';
import styles from './ProductExtract.module.css';
import { HEADING } from '~/constants';

import { Heading, Hyperlink } from '~/components';

const ProductExtract = forwardRef(function ProductExtractRef(
  {
    className,
    dataTestRef,
    hasBottomBorder,
    isVisible,
    product,
    theme,
    itemNum,
  },
  ref,
) {
  const classSet = cx(
    styles.base,
    styles[theme],
    {
      [styles.hasBottomBorder]: hasBottomBorder,
      [styles.slideIn]: isVisible,
    },
    styles[`item${itemNum}`],
    className,
  );

  const linkClassSet = cx(styles.link);

  const headingClassSet = cx(
    styles.heading,
    {
      [styles.slideIn]: isVisible,
    },
    className,
  );

  const productNameClassSet = cx(
    styles.productName,
    {
      [styles.slideIn]: isVisible,
    },
    className,
  );

  const productImageClassSet = cx(
    styles.productImage,
    {
      [styles.slideIn]: isVisible,
    },
    className,
  );

  const isEmptyObject = value => {
    return Object.keys(value).length === 0 && value.constructor === Object;
  };

  if (!product || isEmptyObject(product)) {
    return null;
  }

  return (
    <Fragment>
      <div className={classSet} ref={ref}>
        <Hyperlink
          className={linkClassSet}
          dataTestRef={dataTestRef}
          title={product.name}
          url={product.url}
        >
          <Heading
            className={headingClassSet}
            level={HEADING.LEVEL.FOUR}
            size={HEADING.SIZE.X_X_SMALL}
          >
            <span>{product.works}</span>
          </Heading>
          <div className={productNameClassSet}>
            <span>{product.name}</span>
          </div>
          <div className={productImageClassSet}>
            <Image
              altText={product.image.altText}
              small={product.image.small}
            />
          </div>
        </Hyperlink>
      </div>
    </Fragment>
  );
});

ProductExtract.propTypes = {
  className: PropTypes.string,
  dataTestRef: PropTypes.string.isRequired,
  hasBottomBorder: PropTypes.bool,
  isVisible: PropTypes.bool,
  itemNum: PropTypes.number,
  product: PropTypes.shape({
    works: PropTypes.string,
    image: PropTypes.object,
    name: PropTypes.string,
    url: PropTypes.string,
  }),
  theme: PropTypes.oneOf(['dark', 'light']),
};

ProductExtract.defaultProps = {
  className: undefined,
  dataTestRef: undefined,
  hasBottomBorder: false,
  isVisible: true,
  itemNum: 1,
  product: {
    works: 'Works well with',
    image: {
      altText: 'Product Extract',
    },
  },
  theme: 'dark',
  url: '/',
};

export default ProductExtract;
