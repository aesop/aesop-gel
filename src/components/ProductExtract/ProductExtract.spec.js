import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import ProductExtract from './ProductExtract';
import ProductExtractFixture from './ProductExtractfixture';

configure({ adapter: new Adapter() });

describe('<ProductExtract />', () => {
  it('should be defined', () => {
    expect(ProductExtract).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <ProductExtract
          dataTestRef={ProductExtractFixture.dataTestRef}
          product={ProductExtractFixture.product}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('returns null product prop is not a populated object', () => {
    const component = shallow(
      <ProductExtract
        dataTestRef={ProductExtractFixture.dataTestRef}
        product={{}}
      />,
    );
    expect(component.type()).toEqual(null);
  });
});
