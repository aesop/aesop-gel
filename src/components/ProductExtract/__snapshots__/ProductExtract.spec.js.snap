// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductExtract /> renders base component correctly 1`] = `
<div
  className="base dark slideIn item1"
>
  <a
    className="base left dark link"
    data-test-ref="test_id"
    href="/"
    target="_self"
    title="The sea"
  >
    <div
      className="flexContainer delayAnimation slideIn"
    >
      <div>
        <h4
          className="base xXSmall dark heading delayAnimation slideIn"
        >
          <span>
            Works well with
          </span>
        </h4>
        <div
          className="productName delayAnimation slideIn"
        >
          <div>
            <span>
              The sea
            </span>
          </div>
        </div>
      </div>
      <div
        className="productImage delayAnimation slideIn"
      >
        <picture
          className="base picture dark"
        >
          <source
            media="(min-width: 1025px)"
            srcSet="https://images.ctfassets.net/u1nb1km7t5q7/5JHOiuwRLBAc78jTcLVb5I/bfad1667393889dd8af5e31b080436bd/Aesop_Hair_Violet_Leaf_Hair_Balm_60mL_small.png"
          />
          <source
            media="(min-width: 640px)"
            srcSet="https://images.ctfassets.net/u1nb1km7t5q7/5JHOiuwRLBAc78jTcLVb5I/bfad1667393889dd8af5e31b080436bd/Aesop_Hair_Violet_Leaf_Hair_Balm_60mL_small.png"
          />
          <source
            media="(min-width: 0px)"
            srcSet="https://images.ctfassets.net/u1nb1km7t5q7/5JHOiuwRLBAc78jTcLVb5I/bfad1667393889dd8af5e31b080436bd/Aesop_Hair_Violet_Leaf_Hair_Balm_60mL_small.png"
          />
          <img
            alt="Product Extract"
          />
        </picture>
      </div>
    </div>
  </a>
</div>
`;
